### PR TITLE
The gradient is not flowing 

### PR DESCRIPTION
**Pull request by amelie-girard7**
Hi Inigo,

We're having a problem with the score_embeds function. the gradients aren’t flowing. Mind taking a look when you get a chance? I thought the code I used before was the same as yours, so not sure what’s going on.

Thanks!

### CODE REVIEW

File: src/dto/models/model.py [hunk 1]
```diff
@@ -60,166 +119,266 @@ def train(self, mode=True):
...
-        probs = torch.softmax(logits, dim=-1)
 
-        # Get the embedding matrix and explicitly move it to the logits' device
+        # ===== 3. Debug Prints =====
+        print("\n>> Forward Pass")
+        print(f"Logits: {logits.shape} (device: {device})")
+        print(f"Range: [{logits.min().item():.2f}, {logits.max().item():.2f}]")
+
+        # ===== 4. Gumbel-Softmax =====
+        probs = self.gumbel_softmax(logits)
```
Comment (inigo-jauregi-rozetta) [inline src/dto/models/model.py:141]: @amelie-girard7 Where is the `self.gumbel_softmax` private function defined? I cannot see it in the constructor. I only see the configuration parameters, but not the function itself.

File: src/dto/models/model.py [hunk 2]
```diff
@@ -60,166 +119,266 @@ def train(self, mode=True):
...
-        generated_tokens = self.model.generate(
-            input_ids=input_ids,
-            attention_mask=attention_mask
-        )
-        print("Generated tokens size:", generated_tokens.size())
-
-        # Decode the generated tokens to obtain text
-        generated_texts = self.tokenizer.batch_decode(
-            generated_tokens, skip_special_tokens=True
+        outputs = self.model(  # Store the full outputs to access logits later
```
Comment (inigo-jauregi-rozetta) [inline src/dto/models/model.py:260]: @amelie-girard7 
This does not need to be called during training. We only need to calculate the expected embeddings. So you should call the `self.forward` function, like you are doing it below, but not the model.

File: src/dto/models/model.py [hunk 3]
```diff
@@ -60,166 +119,266 @@ def train(self, mode=True):
...
+            prog_bar={'train/dto_loss': True},
+            logger=True
         )
+        
+        return dto_loss_val
+
+    def validation_step(self, batch, batch_idx):
+        with torch.no_grad():  # CRITICAL - no gradients during validation
+            # Forward pass
+            expected_embeddings = self.forward(
```
Comment (inigo-jauregi-rozetta) [inline src/dto/models/model.py:323]: @amelie-girard7 
Usually the validation evaluation metrics are identical to the test, so probably we should not use the expected embeddings here. Instead we should obtain the hard predictions. Maybe you could use the `self.model` function directly, or the `generate` function like in the test set. 

Or are you aiming to select the best checkpoint based on the dto loss? Then it is ok to do what you are doing. Does the validation loss decrease over time?
